### ISSUE TITLE
crash upon setting alignment

### ISSUE THREAD

**Issue by Boruch-Baum** [blippy/neoleo#39]
Hi. I just found your version of `neoleo` (v8.0.0), and am beginning to go through the documentation; thanks for the work!

I've downloaded the tarball to a debian environment, and performed `make`, but not `make install`. By running `./neoleo` from the `src` directory, the program loads and can perform a basic calculation of adding two cells; however, when attempting to set alignment using either the `C-l` or `C-r` keybinding, the program crashes, with an error message `free(): invalid pointer`.

BTW: I have other things to report (basic keybindings and commands not functioning), but thought to first post this, and get feedback.

**Comment by blippy** [comment]
Ah yes, C-r. I'm aware of the crash. The problem is: what to do, what to do? The project inherits a lot of legacy C code, and it's difficult to make fit in with C++. Not to mention that the code is ferociously complex in places. Changing a piece of code can lead to all sorts of cascading compile errors.

The code is often difficult to reason about. Sometimes memory is copied, sometimes only the pointer is copied, sometimes a memory pool is used. The `free()` problem you reported is the result of two different ways a variable called `expanded_prompt` is set. 

I'm actually thinking that the way that Neoleo currently handles commands is too complex, and a switch from emacs-like keybindings to vi-like keybindings is a better bet ... ultimately.

I'll take a look at the `free()` problem you mention.

What I will ask is: at this stage, only worry about functionality you do use, not the functionality you don't. There are a lot of problems with the code, with some basic structure needing to be sorted out. That's why I haven't been shy about removing functionality I don't use. It's a complex beast, and all of its problems can't be resolved quickly. It's a marathon, rather than a sprint.

Hope that helps.

**Comment by Boruch-Baum** [comment]
Understood.

Do you want me to continue reported bugs that I encounter (bad keybindings / commands)? If so, should I do so as a continuation of this thread?

FWIW - I prefer emacs-style over vi.

Also, I've downloaded version 7.0.0, and can compare functionality in future reports. The indentation commands do function in v7

**Comment by blippy** [comment]
It seems that github automatically closed this issue, which wasn't my intention. But it does fix the C-l C-r issue.

I'm surprised that the alignment commands work in v7.0.0. Hmm, very strange.

As regards bad keybindings/commands, just report the issues that are meaningful for you. Please report them on different threads, though. A thorough audit of the functionality would reveal a lot of stuff that doesn't work as advertised, so it's best that we're selective. 

I'm keen to eliminate as much unneeded functionality from Neoleo as possible. The original GNU version tried to be a professional spreadsheet, with things like Motif interfaces, and suchlike. It stagnated, and was overtaken by Gnumeric and LibreOffice. So I have taken a different tack: try to make things as simple as possible, keep the dependencies low, and stick to an ncurses interface. That way, one has an interface that works over a remote terminal. Or maybe one just doesn't want to use X-Windows.

One thing Neoleo does have is a command-line mode (aka "headless"), which was one of the best decisions I made. It allows me to test the spreadsheet via scripts, and use UNIX pipelining. In that way, it's actually a bit of a win over traditional spreadsheets, which try to be applications.

**Comment by Boruch-Baum** [comment]
it didn't occur to me that the original neoleo had wanted to be an "all-environment" spreadsheet - if I wanted X11 or Motif, I would choose something else.

OTOH, when it comes to actual functionality of the spreadsheet, by that I mean the list of embedded functions and the user-friendliness of the interface - I hope that isn't in your plans to strip.

If I have time later today, I'll take a look at the code and see how its organized or dis-organized as the case may be.

Also, I'll open another thread.

**Comment by blippy** [comment]
> it didn't occur to me that the original neoleo had wanted to be an "all-environment" spreadsheet - if I wanted X11 or Motif, I would choose something else.

Well, they seemed quite ambitious. It's a very old project, so it is likely to be one of the first open-source spreadsheets there is. Someone had started to put a GTk into it, although it was only a couple of functions, so that it never would have worked as-is.

The X11 version was OK, I guess, but didn't seem to add too much. They'd put a lot of effort into the Motif one, but I decided to rip it out. In retrospect, this was a good idea. Motif is quite old, which although I didn't so much because hey, a bit of retro doesn't go amiss, it was actually quite limited. It relied on a component called Xbae. There are now conflicting implementations of Motif. There's LessTif, which is a free version that people created because the original one was proprietary. They've now open-sourced it, which is actually a bit of a drag, because it can create package conflicts. 

I decided that I was better off without Motif. And, as you note, if you want something for X11, there are choices in the form of gnumeric and LibreOffice. Gnumeric seems quite lightweight to me. If I were to make it available on X-Windows, I'd probably choose something like FLTK. I'm now pretty sure I want to keep it a console-only app.
 
> OTOH, when it comes to actual functionality of the spreadsheet, by that I mean the list of embedded functions and the user-friendliness of the interface - I hope that isn't in your plans to strip.

I've stripped some functions from it, although if you really need them, I can write them back in.

I'm surprised to hear you say that the spreadsheet is user-friendly. I've read a few reviews on oleo, and people have found the interface to be somewhat puzzling. Knowing emacs definitely helps, though.


> If I have time later today, I'll take a look at the code and see how its organized or dis-organized as the case may be.

There's some odd things going on. There's a lot of cutting-and-pasting going on. Not that i should be too judgemental, of course, because i've written a fair amount of garbage in it, too. 

They also have this thing where they they run a header file through the C preprocessor a few times in order to generate the right bits for the right part of the code. Urggh!

I've run the program `complexity` on the source code, and the upshot is: yes, it's complex!

**Comment by Boruch-Baum** [comment]
I didn't say that the spreadsheet was user-friendly. I was trying to convey that stripping pre-existing user-friendly features is undesirable. For instance, I see that version 7 had a kind of integrated help system via C-h, that isn't in version 8, and seems nice. On the other hand, please keep in mind that I've only spent on the order of minutes using the program, so I'm not at this point a well-informed user or a committed stakeholder.

I see that there exists in the debian repositories a package called `sc`, which is described as a "very stable" command-line spreadsheet with vi keybindings. Are you familiar with that project? How does it compare to this one?

**Comment by blippy** [comment]
I am vaguely aware of `sc`, and used it in the past. As far as I know, `sc` is stable, and has a long lineage. There's a program called `sc-im` which is a fork, and improves on it in many ways, I believe.  It seems to have a lot of cools stuff, like undo, coloured text, and multiple sheets. So it looks a very good project.

I took a look at the source code of sc-im. It seemed pretty good, but I concluded that I didn't want to jump ship from oleo. neoleo started out as just a little hobby of mine as using something from the dim-and-distant past. I had an Amiga in the 90's, and I think that oleo worked on it. I'm not sure, though.

One thing neoleo does have is an `IF` statement, which I don't think `sc-im` has. I actually use `IF` statements in some spreadsheet. I'm not sure how `sc-im` handles dates, either. There was a guy who said that he liked neoleo because of the way it handled dates. Other spreadsheets didn't do quite what he wanted.

So `sc`/`sc-im` are quite capable programs, readily avaliable, and may suit your taste more. `neoleo` flies under most people's radar, which has its advantages in that expectations are low.
